doc: add change doc to specify max plaintext length input on encrypt (#138)
awslabs/aws-encryption-sdk-specification: changes/2020-06-17_encrypt-max-plaintext-length-input/change.md

[//]: # "Copyright Amazon.com Inc. or its affiliates. All Rights Reserved."
[//]: # "SPDX-License-Identifier: CC-BY-SA-4.0"

# Encrypt Max Plaintext Length Input

## Affected Features

| Feature                                           |
| ------------------------------------------------- |
| [Encrypt](../../client-apis/encrypt.md)           |
| [CMM Interface](../../framework/cmm-interface.md) |

## Affected Specifications

| Specification                                     |
| ------------------------------------------------- |
| [Encrypt](../../client-apis/encrypt.md)           |
| [CMM Interface](../../framework/cmm-interface.md) |

## Affected Implementations

| Language   | Repository                                                                            |
| ---------- | ------------------------------------------------------------------------------------- |

> | Python     | [aws-encryption-sdk-c](https://github.com/aws/aws-encryption-sdk-python)              |

Reviewer: I think this label should read "aws-encryption-sdk-python".

| Javascript | [aws-encryption-sdk-javascript](https://github.com/aws/aws-encryption-sdk-javascript) |

## Definitions

### Conventions used in this document

The key words
"MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
"SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
in this document are to be interpreted as described in
[RFC 2119](https://tools.ietf.org/html/rfc2119).

### Known-Length Plaintext

Any plaintext input to the Encrypt where the total length of that plaintext
can be immediately determined.
For example, if a user supplies plaintext as a string or byte array to Encrypt,
that is a known-length plaintext.

### Unknown-Length Plaintext

Any plaintext input to the Encrypt operation where the total length of that plaintext
cannot be immediately determined.
For example, if a user supplies plaintext as a stream to Encrypt,
that is a unknown-length plaintext.

## Summary

When performing the encrypt operation on an unknown length plaintext,

> users MUST be able to specify some input that bounds the length of the plaintext to be encrypted.

Reviewer: Should the summary also state that users MUST NOT be able to specify the maximum plaintext length when performing Encrypt on a known-length plaintext?

Author: We need to be careful about the wording, but yes. I've updated the summary.

This input already exists in the specification as [plaintext length](../../client-apis/encrypt.md#plaintext-length)

> however it's effect on the Encrypt operation's behavior is unspecified.

Reviewer: nit: its

This change renames this input in the specification

> to be a more accurate reflection of it's behavior,

Reviewer: nit: its

and specifies the correct behavior during the Encrypt operation.

## Out of Scope

- How different implementations may allow a user to specify this input for the Encrypt operation
  is out of scope.
- Significantly changing the shape of any of our Encrypt API implementations is out of scope.

## Motivation

All implementations allow users to specify a plaintext length when calling Encrypt with a
unknown-length plaintext.

> However, it's behavior in the Encrypt operation is not clearly specified.

Reviewer: nit: its

Additionally, some implementations also allow a plaintext length to be specified on Encrypt with a
known-length plaintext.
The behavior of Encrypt in this case is not consistent between implementations.

The purpose of this change is to document the correct behavior of this input and

> rename this input to clarify it's intent.

Reviewer: nit: its

then such a CMM is not correctly using the intent behind `maxPlaintextLength`.
It is better to restrict flexibilty here so that users have less opportunity
to depend on behaviors from CMMs that are misusing this value.

## Drawbacks

This change does not prescribe the name of this input in implementations,
and thus there will not be a standardized name or way of specifying this input across implementations.
This is because this control already exists in implementations under different names,
and it is not worth breaking all customers to rename this input
without significantly changing this input to benefit the customer.

## Security Implications

This change SHOULD NOT have any security implications.

## Operational Implications

This change will break any JS user that is currently using Encrypt with a known-length plaintext
and inputting a plaintextLength less than that length.

This change will break any Python user that depends on the source_length being sent to the
CMM instead of the true length of a known-length plaintext.

## Guide-level Explanation

This change renames the `plaintext length` Encryption input to `max plaintext length` in the specification,

> and clarifies it's behavior in the Encrypt operation.

Reviewer: nit: its

If during the Encrypt operation it is determined that
the actual plaintext length is greater than a supplied `max plaintext length`,
the operation MUST fail.

Additionally, when calling the CMM's GetEncryptionMaterials call during the Encrypt operation,
the operation MUST pass in the length of the input plaintext if it can be known.
If it can't be known, the operation passes in `max plaintext input` if supplied.
If it can't be known and no `max plaintext input` is supplied, the operation supplies no value for
this field.

This change also renames the `plaintext length` field in the CMM interface to `max plaintext length`.

Finally, this change specifies when this input should be available for users to specify:
if and only if the input plaintext can be of unknown-length.

## Reference-level Explanation

We need to update the spec to make the purpose and behavior of the `plaintext length` input clear

> by renaming it and specifying it's exact behavior within the Encrypt operation:

Reviewer: nit: its

known-length plaintext.
The behavior of Encrypt in this case is not consistent between implementations.

The purpose of this change is to document the correct behavior of this input and
rename this input to clarify its intent.

This change renames this input within the spec because
the intention behind the value of this input is that
it is a bound on the plaintext length,
not the exact value.
This is what the Encrypt operation enforces when this value is set,
and this is how CMMs should interpret this value in GetEncryptionMaterials if it is set.
Implementations already treat this input with this intent in mind,
so we should describe it in the specification as such.

This change specifies that the Encrypt operation MUST pass the real length of the plaintext
to the CMM if known instead of a user supplied bound because the former is more accurate.
There is no case where a user would benefit from passing
a less accurate `max plaintext length` to the CMM.
If there were such a case,
then such a CMM is not correctly using the intent behind `maxPlaintextLength`.
It is better to restrict flexibilty here so that users have less opportunity

> to depend on behaviors from CMMs that are misusing this value.

Reviewer 2: To be clear: Using this behavior to try and flush the Caching CMM is not a good idea.
Say, trying to encrypt 0 bytes with a plaintext length of 2**66 to ensure
that the next encrypt gets a new data key.

Author: That is a more concrete example of what I'm trying to avoid. Users should not be using this value to be triggering weird behaviors in the CMM. Maybe framing as "the CMM is doing a bad thing" isn't the correct way to frame it though, and instead say "users shouldn't expect setting this value to change the CMM's behavior".

## Drawbacks

This change does not prescribe the name of this input in implementations,
and thus there will not be a standardized name or way of specifying this input across implementations.
This is because this control already exists in implementations under different names,
and it is not worth breaking all customers to rename this input
without significantly changing this input to benefit the customer.

## Security Implications

This change SHOULD NOT have any security implications.

## Operational Implications

This change will break any JS user that is currently using Encrypt with a known-length plaintext

> and inputting a plaintextLength less than that length.

Reviewer 2: No, the 2 JS implementations do not let this happen.
The browser does not accept the option at all because **all** operations are one-shot.
https://github.com/aws/aws-encryption-sdk-javascript/blob/master/modules/encrypt-browser/src/encrypt.ts#L40
The Node implementation will clobber any value provided in favor of the correct value. 
https://github.com/aws/aws-encryption-sdk-javascript/blob/master/modules/encrypt-node/src/encrypt.ts#L30

Author: I will update to specify "Node JS". The Node JS implementation violates the spec I've outlined here. Instead of clobbering in the "less than" case, it MUST fail. Therefore, anyone who was previously depending on the implementation just using the correct value in the "less than" case will break.

Reviewer 2: Hum, if what you want is to "not let it be an option" then how would that work, if not this way?

This change will break any Python user that depends on the source_length being sent to the
CMM instead of the true length of a known-length plaintext.

## Guide-level Explanation

This change renames the `plaintext length` Encryption input to `max plaintext length` in the specification,
and clarifies its behavior in the Encrypt operation.

If during the Encrypt operation it is determined that
the actual plaintext length is greater than a supplied `max plaintext length`,
the operation MUST fail.

Additionally, when calling the CMM's GetEncryptionMaterials call during the Encrypt operation,
the operation MUST pass in the length of the input plaintext if it can be known.
If it can't be known, the operation passes in `max plaintext input` if supplied.
If it can't be known and no `max plaintext input` is supplied, the operation supplies no value for
this field.

This change also renames the `plaintext length` field in the CMM interface to `max plaintext length`.

Finally, this change specifies when this input should be available for users to specify:

> if and only if the input plaintext can be of unknown-length.

Reviewer 2: How should this be accomplished?
The JS encrypt function takes multiple options.
JavaScript does not have any function overloads.

Should there be multiple `encrypt` functions?
What should they be called?
Or is what it is doing currently adequate?

What does "available" mean?

Author: I think answers to these questions are more clearly addressed in the "Reference-level Explanation" section. I was actually a bit unsure what detail to go into in this one. As it is right now, this sentence is not intending to specify what the exact shape of the API should look like, only that if you provide an API that "only" takes in known-length plaintexts, then it doesn't make sense for users to be able to specify this input.

## Reference-level Explanation

We need to update the spec to make the purpose and behavior of the `plaintext length` input clear
by renaming it and specifying its exact behavior within the Encrypt operation:

When performing the encrypt operation on a unknown-length plaintext,
users MUST be able to specify an optional parameter `maxPlaintextLength`.
The value of this field represents the bound on the length of the plaintext to be encrypted.
The ESDK MUST NOT encrypt a plaintext greater than this length,
and MUST fail if it can be determined during encryption that the actual plaintext length
is greater than what the user supplied on input.
The actual name of this input, and how the user specifies this value for the Encrypt operation
MAY be different per implementation.

> We also need to specify the exact behavior for how this input is used in the CMM's GetEncryptionMaterials call:

Reviewer 2: ```suggestion
We also need to specify the exact behavior for how this input is passed to the CMM's GetEncryptionMaterials call:
```

I don't think you mean "used" because you don't talk about how the CMM is using the value.

| Python   | [aws-encryption-sdk-python](https://github.com/aws/aws-encryption-sdk-python) |

## Definitions

### Conventions used in this document

The key words
"MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
"SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
in this document are to be interpreted as described in
[RFC 2119](https://tools.ietf.org/html/rfc2119).

### Known-Length Plaintext

Any plaintext input to the Encrypt operation where the total length of that plaintext
can be immediately determined.
For example, if a user supplies plaintext as a string or byte array to Encrypt,
that is a known-length plaintext.

### Unknown-Length Plaintext

Any plaintext input to the Encrypt operation where the total length of that plaintext
cannot be immediately determined.
For example, if a user supplies plaintext as a stream to Encrypt,
that is a unknown-length plaintext.

## Summary

The Encrypt operation takes a plaintext as input that can either have a known length or an
unknown length.
If the input plaintext is a unknown-length plaintext,
the Encrypt operation MUST also take an optional input that bounds the length of that input.

> The specification already specifies a [plaintext length](../../client-apis/encrypt.md#plaintext-length)

Reviewer 3: Should this be an absolute link (with the SHA)? So a different header in the future still works for the link?

input on the Encrypt operation that describes a bound on the plaintext length,
however its effect on the Encrypt operation's behavior is unspecified.

This change renames this input to `plaintext length bound` in the specification
to be a more accurate reflection of its behavior and intent,
and specifies the correct behavior during the Encrypt operation.

> Specifically, in the scope of the Encrypt operation it's value MUST be passed to the CMM.

Reviewer 3: np: `its`

Additionally, some implementations also allow a plaintext length to be specified on Encrypt with a
known-length plaintext.
The behavior of Encrypt in this case is not consistent between implementations.

The purpose of this change is to specify exactly how this input affects the Encrypt operation.

This change renames this input to `plaintext length bound` within the spec because
the intention behind the value of this input is that
it is a max bound of the plaintext length,
set by the user when streaming encryption.
If the actual plaintext length is greater than this value,
something is wrong and the operation MUST fail.
Otherwise, the Encrypt operation MUST pass this value to the CMM.

The intention of this value is not to be a one to one passthrough value to the CMM.
If that were the case,
then we would be concerned with always letting users specify this value for known-length plaintexts,
or with possibly letting users pass in a "unknown length" intent.
We do not want to support these cases.
Users should not be setting a `max plaintext length` value for known-length plaintexts
with the intention of changing CMM behavior.

> Instead, users should be specifying this value in for unknown-length plaintexts in order to use CMMs

Reviewer 3: Users `should`/`should not` or users `must`/`must not`?

Author: The first is definitely a SHOULD NOT, which makes the latter be a SHOULD. Alternatively, I'd be ok with using a different verb to avoid RFC 2119 verbs when not strictly necessary.

that depend on plaintext length for their internal logic.

This change additionally renames the `plaintext length` field on GetEncryptionMaterialsRequest to
`max plaintext length` because that better explains the intent behind this field.
This field describes the max size of the plaintext which will be encrypted using the materials
the CMM is to generate.
This is the current state of the behavior, so the specification should describe it as such.
This is renamed to a different name than `plaintext length bound` in order to
distinguish the two controls.

This change does not strictly prescribe how users should express intent for the plaintext
and a possible `plaintext length bound` through an API.
How an API should be designed with these controls in mind depends greatly on the capabilities and
idioms of the language in use.
Specifically, this gets tricky for languages where the idiom is to

> accept an object represent a set of user supplied params.

Reviewer 3: accept an object that represents a set?

If a user supplies both a known-length plaintext and a `max plaintext length`,
what should be done?
It doesn't make sense to have to check for bad values in the user supplied params,
given it's unstructured nature.

> Thus, our specification reccomends that languages SHOULD either ensure that `max plaintext length`

Reviewer 3: np: `recommends`

np: `languages SHOULD either ensure ` -> `languages SHOULD ensure ` ?

isn't supplied with a known-length plaintext by construction.
If it doesn't, then it SHOULD ignore any user supplied `max plaintext length` in the case that
a known-length plaintext is supplied.

## Drawbacks

This change does not prescribe the name of this input in implementations,
and thus there will not be a standardized name or way of specifying this input across implementations.
This is because this control already exists in implementations under different names,

> and it is not worth breaking all customers to rename this input

Reviewer 3: np: `breaking all customers` -> `breaking customer-facing APIs`

without significantly changing this input to benefit the customer.

## Security Implications

This change SHOULD NOT have any security implications.

## Operational Implications

This change will break any Python user that depends on the source_length being sent to the
CMM instead of the true length of a known-length plaintext.

## Guide-level Explanation

We need to update the spec to make the purpose and behavior of the `plaintext length` input clear
by renaming it and specifying its exact behavior within the Encrypt operation:

When performing the Encrypt operation on an unknown-length plaintext,
users MUST be able to specify an optional parameter `plaintext length bound`.
The value of this field represents the max length of the plaintext to be encrypted.

> The ESDK MUST NOT encrypt a plaintext greater than this length,

Reviewer 2: I feel like this is a little buried.  I can see why it is not in the list below, but I think that a little higher billing is in order.

Author: Where do you think this should be highlighted more? In the summary? Reference it again later?

a known-length plaintext is supplied.

## Drawbacks

This change does not prescribe the name of this input in implementations,
and thus there will not be a standardized name or way of specifying this input across implementations.
This is because this control already exists in implementations under different names,
and it is not worth breaking customer facing APIs to rename this input
without significantly changing this input to benefit the customer.

## Security Implications

This change SHOULD NOT have any security implications.

## Operational Implications

This change will break any Python customer that depends on the source_length being sent to the
CMM instead of the true length of a known-length plaintext.

## Guide-level Explanation

We need to update the spec to make the purpose and behavior of the `plaintext length` input clear
by renaming it and specifying its exact behavior within the Encrypt operation:

When performing the Encrypt operation on an unknown-length plaintext,
customers MUST be able to specify an optional parameter `plaintext length bound`.
The value of this field represents the max length of the plaintext to be encrypted.

> The ESDK MUST NOT encrypt a plaintext greater than this length,
> and MUST fail if it can be determined during encryption that the actual plaintext length
> is greater than what the customer supplied on input.

Reviewer 2: ```suggestion
The ESDK MUST NOT encrypt a plaintext greater than this length.
If it is determined during encryption that the actual plaintext length
is greater than what the customer supplied on input
the ESDK MUST fail.
```

If the input plaintext is a unknown-length plaintext,
the Encrypt operation MUST also take an optional input that bounds the length of that input.

The specification already specifies a
[plaintext length](https://github.com/awslabs/aws-encryption-sdk-specification/blob/be6870b2513f4ca44bea8f2e0b5eb4808dba4365/client-apis/encrypt.md#plaintext)
input on the Encrypt operation that describes a bound on the plaintext length,
however its effect on the Encrypt operation's behavior is unspecified.

This change renames this input to `plaintext length bound` in the specification
to be a more accurate reflection of its behavior and intent,
and specifies the correct behavior during the Encrypt operation.
Specifically, in the scope of the Encrypt operation its value MUST be passed to the CMM.

Additionally, it is unclear in the spec what the behavior should be
if the input plaintext has a known-length
and the customer specifies a bound on the length of that plaintext.

This change proposes that in the scope of the Encrypt operation,
an input plaintext is either of known length,
of unknown length with customer supplied estimated size,
or unknown length with no customer supplied estimated size.

> The Encrypt operation should never consider a `plaintext length bound` alongside

Reviewer 4: should not consider or should not allow?

The purpose of this change is to specify exactly how this input affects the Encrypt operation.

This change renames this input to `plaintext length bound` within the spec because
the intention behind the value of this input is that
it is a max bound of the plaintext length,
set by the customer when streaming encryption.
If the actual plaintext length is greater than this value,
something is wrong and the operation MUST fail.
Otherwise, the Encrypt operation MUST pass this value to the CMM.

The intention of this value is not to be a one to one passthrough value to the CMM.
If that were the case,
then we would be concerned with always letting customers specify this value for known-length plaintexts,
or with possibly letting customers pass in a "unknown length" intent.
We do not want to support these cases.
Customers SHOULD NOT set a `max plaintext length` value for known-length plaintexts
with the intention of changing CMM behavior.
Instead, customers SHOULD specify this value for unknown-length plaintexts in order to use CMMs
that depend on plaintext length for their internal logic.

This change additionally renames the `plaintext length` field on GetEncryptionMaterialsRequest to

> `max plaintext length` because that better explains the intent behind this field.

Reviewer 4: I've read this through several times now and I just get more confused each time. :/

What is the difference between `plaintext length bound` and `max plaintext length` other than one being part of the Encrypt API and one being part of the CMM GetEncryptionMaterials API?

Author: `max plaintext length` is an input on the CMM GetEncryptionMaterials API. It describes the maximum size of the plaintext that the CMM caller intends to encrypt using the result materials. I assert that in the context of Encrypt, if the plaintext has a known length this MUST be exactly the length of the plaintext.

`plaintext length bound` is an input on Encrypt. If provided, it describes the maximum size of a unknown-length plaintext. It specifically MUST NOT describe the intent of "I, as a caller, want the Encrypt operation to pass this value in to the CMM instead of the actual length."

They both describe "plaintext MUST NOT be larger than this size", but the intentions are subtly different. In order to distinguish them, I have given them slightly different names.

> This field describes the max size of the plaintext which will be encrypted using the materials

Reviewer 4: ```suggestion
This field describes the max size of the plaintext that will be encrypted using the materials
```

the CMM is to generate.
This is the current state of the behavior, so the specification should describe it as such.
This is renamed to a different name than `plaintext length bound` in order to
distinguish the two controls.

This change does not strictly prescribe how customers should express intent for the plaintext
and a possible `plaintext length bound` through an API.
How an API should be designed with these controls in mind depends greatly on the capabilities and
idioms of the language in use.
Specifically, this gets tricky for languages where the idiom is to
accept an object that represents a set of customer supplied params.
If a customer supplies both a known-length plaintext and a `max plaintext length`,
what should be done?
It doesn't make sense to have to check for bad values in the customer supplied params,
given its unstructured nature.
Thus, our specification recommends that languages SHOULD ensure that `max plaintext length`
isn't supplied with a known-length plaintext by construction.
If it doesn't, then it SHOULD ignore any customer supplied `max plaintext length` in the case that

> a known-length plaintext is supplied.

Reviewer: Assuming the guide-level explanation below is correct, then this should be a MUST rather than a SHOULD.
